### ISSUE TITLE
forceRefreshSession prematurely completes its observable

### ISSUE THREAD

**Issue by PartyArk** [damienbod/angular-auth-oidc-client#767]
I'm having difficultly subscribing to forceRefreshSession with consistent results - here's what I _think_ is happening -

1. `forceRefreshSession` observes `refreshSessionWithIFrameCompleted$`
2. `refreshSessionWithIFrameCompleted$` is internally the Subject for `silentRenewEventHandler`
3. The Iframe does it's thing, and `silentRenewEventHandler` is triggered - it fires a status back to `refreshSessionWithIFrameCompleted$` via `fireRefreshWithIframeCompleted` (all this works as expected)
4.  This is where I think the problem is - although `fireRefreshWithIframeCompleted` correctly updates the internal `refreshSessionWithIFrameCompleted$` Subject, it then sets the Subject to **complete** 
5. So, every **subsequent** time you do `forceRefreshSession`, even though it works (ie refreshes the session), you don't get anything back to observe, because, even though `fireRefreshWithIframeCompleted` tries to set the Subject value, it's already **complete** and so doesn't do anything.

It's a bit obscure - I picked it up because I'm doing a session refresh immediately when a browser tab comes back into focus. Thanks.

**Comment by damienbod** [comment]
@PartyArk We have a bug here, which we fix now for the next version. Thanks for reporting

Greetings Damien

**Comment by PartyArk** [comment]
Hi thanks for taking a look - just had a look at the branch.

I thin you may want to stick to using a regular Subject for the `refreshSessionWithIFrameCompletedInternal$` in the SilentRenewService; a BehaviorSubject will immediately return its initial value of null on observation.

To simplify a lot of filtering on the RefreshSessionService `forceRefreshSession`  merge or forkjoin (and make it simpler)  I would suggest doing the refreshSession and callbackContext consecutively, something like

```
    const callBackContext$ = this.silentRenewService.refreshSessionWithIFrameCompleted$;

    return this.startRefreshSession().pipe(
      switchMap(() => callBackContext$),
      map((callbackContext) => {
        const isAuthenticated = this.authStateService.areAuthStorageTokensValid();
        if (isAuthenticated) {
          return {
            idToken: callbackContext?.authResult?.id_token,
            accessToken: callbackContext?.authResult?.access_token,
          };
        }
        return null;
      })
    );

```

This would also make it very similar to the code-flow-with-refreshTokens method.

Thanks.

**Comment by PartyArk** [comment]
Hi thanks for taking a look at this; I jumped the gun a bit earlier, but now have had a look at your fix.

I tried it out but there are a couple of issues I found

1. If silent renew returns an error, the callbackContext can return a result _before_ the startRefreshSession is completed. So you can't wait for the first to complete. If you do that, you can have a scenario where you'll never get a result.

2. Because there is a BehaviorSubject, you can get back null when you first subscribe (e.g. load a new tab), which is an error - not sure what the thinking behind changing it is.

Here's what I did -

1. set the refreshSessionWithIFrameCompletedInternal$ back to a Subject (not BehaviorSubject)

2. the key routine in forceRefreshSession looks like this -

```

      // get the first results from these two
      const refreshSession$ = this.startRefreshSession().pipe(first());
      const callBackContext$ = this.silentRenewService.refreshSessionWithIFrameCompleted$.pipe(first());

      return forkJoin(refreshSession$, callBackContext$).pipe(
        map(([sessionResult, callbackContext]) => {

          // ensure session setup
          if (sessionResult !== true) { return null; }

          // indicates an error in the callback
          if (callbackContext === null) { return null; }

          const isAuthenticated = callbackService.authStateService.areAuthStorageTokensValid();
          if (isAuthenticated) {
            return {
              idToken: callbackContext?.authResult?.id_token as string,
              accessToken: callbackContext?.authResult?.access_token as string,
            };
          }
          return null; // did not receive valid tokens
        })
      );

```

Hope that helps, I'd do a pull but can't from here.

**Comment by damienbod** [comment]
fixed in version 11.1.3, thanks for reporting
